File: TestTask.Web/src/app/services/category.service.ts
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';

import { ApiService } from './api.service';
import { CategoryDto } from '../models/category';

@Injectable()
export class CategoryService extends ApiService {
    public getList(): Observable<CategoryDto[]> {
        return this.get<CategoryDto[]>('category/list', null);
    }

    public autoComplete(term: string): Observable<CategoryDto[]>{
        return this.get<CategoryDto[]>('category/list', {term: term});
    }
}

File: TestTask.Api/Services/CategoryService.cs
using System.Collections.Generic;
using System.Linq;
using TestTask.Api.DtoBuilders;
using TestTask.Api.Models;
using TestTask.DataAccess.Repositories;

namespace TestTask.Api.Services
{
    public interface ICategoryService
    {
        List<CategoryDto> Autocomplete(string term);
    }

    public class CategoryService:ICategoryService
    {
        private readonly ICategoryRepository _categoryRepository;

        public CategoryService(ICategoryRepository categoryRepository)
        {
            _categoryRepository = categoryRepository;
        }

        public List<CategoryDto> Autocomplete(string term)
        {
            return _categoryRepository.Get(term).Select(x=>CategoryDtoBuilder.ToDto(x)).ToList();
        }
    }
}
